### PR TITLE
Do not inactivate case whenever the assignee is removed from database using the cli

### LINKED ISSUE #2039: Deleting users causes old assigned cases to show as inactive
We suddenly have a bunch of old cases shown in the list of inactive. Most of these were archieved since a while back (several from 2018 and 2019), but might be one or two active cases hiding in there too... 
Have you done any updates or re-run any cases? 

/Sofie

### PR DESCRIPTION
**Pull request by northwestwitch**
fix #2039.

Current situation is that whenever a user is assignee of one or more cases, if the user is removed using the command line, those cases become inactive. This fix will keep those cases in their status, whichever that is.

**How to test --in a local instance of Scout**:
---> From master branch:
1. Create a dummy user to assign the case to:
`scout --demo load user -u "Dummy User" -m dummyuser@dummysite.se -i cust000`
1. Log in into scout with the dummy users's email
1. Enter the case page and then one of its variants; you should get a notification that the case is active.
1. Assign yourself to the case and log out.
1. Remove dummy user from the command line:
`scout --demo delete user -m dummyuser@dummysite.se`
1. Log in as Clark Kent and note that the case is inactivated

--> Repeat from this branch:
1. In the last step you should see that the case is still active

Try to assign whichever state to the case (not inactive of course) to the case before removing the user and the case should keep this status.

**Expected outcome**:
The functionality should be working
Take a screenshot and attach or copy/paste the output.

**Review:**
- [ ] code approved by
- [ ] tests executed by

### CODE REVIEW

**Comment by northwestwitch** [comment]
Ready for review.

**Comment by northwestwitch** [comment]
An easy review, anyone?

**Comment by dnil** [comment]
Hm, I could not reproduce the deactivation when deleting the sole dummy-user from the case?
```
(scout4_36) k1315C02VR0WRHTDH:scout dannil$ scout --demo delete user -m dummy@dummy.se
2020-08-27 14:57:51 k1315c02vr0wrhtdh.gaia.sll.se scout.commands.base[31484] INFO Running scout version 4.20
2020-08-27 14:57:51 k1315c02vr0wrhtdh.gaia.sll.se scout.commands.base[31484] DEBUG Debug logging enabled.
2020-08-27 14:57:51 k1315c02vr0wrhtdh.gaia.sll.se scout.commands.delete.delete_command[31484] INFO Running scout delete user
2020-08-27 14:57:51 k1315c02vr0wrhtdh.gaia.sll.se scout.adapter.mongo.user[31484] INFO Fetching user dummy@dummy.se
2020-08-27 14:57:51 k1315c02vr0wrhtdh.gaia.sll.se scout.adapter.mongo.user[31484] INFO Deleting user dummy@dummy.se
User was correctly removed from database.
2020-08-27 14:57:51 k1315c02vr0wrhtdh.gaia.sll.se scout.adapter.mongo.case[31484] DEBUG Fetch all cases
2020-08-27 14:57:51 k1315c02vr0wrhtdh.gaia.sll.se scout.adapter.mongo.case[31484] INFO Get cases with query {'assignees': {'$in': ['dummy@dummy.se']}}
2020-08-27 14:57:51 k1315c02vr0wrhtdh.gaia.sll.se scout.adapter.mongo.institute[31484] DEBUG Fetch institute cust000
2020-08-27 14:57:51 k1315c02vr0wrhtdh.gaia.sll.se scout.adapter.mongo.case_events[31484] INFO Creating event for unassigning Dummy User from 643594
2020-08-27 14:57:51 k1315c02vr0wrhtdh.gaia.sll.se scout.adapter.mongo.event[31484] DEBUG Saving Event
2020-08-27 14:57:51 k1315c02vr0wrhtdh.gaia.sll.se scout.adapter.mongo.event[31484] DEBUG Event Saved
2020-08-27 14:57:51 k1315c02vr0wrhtdh.gaia.sll.se scout.adapter.mongo.case_events[31484] INFO Updating 643594 to be unassigned with Dummy User
2020-08-27 14:57:51 k1315c02vr0wrhtdh.gaia.sll.se scout.adapter.mongo.case_events[31484] DEBUG Case updated
User was removed as assignee from 1 case(s).
```
But case is still active, although no user is assigned. This is contrary to the test description, but consistent with the last code in the PR. Perhaps there is something left on the local? 😊

**Comment by northwestwitch** [comment]
> This is contrary to the test description

From master it should be inactive, not from this branch

**Comment by dnil** [comment]
I think what you want to add (though you may already have that locally?) is at least an expanded list of exempted states  in adapter.unassign(), e.g. not only 'prioritised' but also 'solved' and 'archived'. Also, in the case of automatic unassigns the same check for other assigned users as for the web interface before requesting an inactivation.

**Comment by northwestwitch** [comment]
> I think what you want to add (though you may already have that locally?) is at least an expanded list of exempted states in adapter.unassign(), e.g. not only 'prioritised' but also 'solved' and 'archived'. Also, in the case of automatic unassigns the same check for other assigned users as for the web interface before requesting an inactivation.

You mean in the pytest test?

**Comment by dnil** [comment]
No, in the actual function. If a single assigned user user is assigned to an active case, and that user is deleted, it still makes sense to inactivate the case. But it becomes highly confusing if that is a case that user solved or archived years back when they were still working actively.

**Comment by dnil** [comment]
The same is true for the web interface. Even if I do not think it is very good practice to mark a case solved, and then unassign, the latter should not really change the case status.

**Comment by northwestwitch** [comment]
> No, in the actual function. If a single assigned user user is assigned to an active case, and that user is deleted, it still makes sense to inactivate the case. But it becomes highly confusing if that is a case that user solved or archived years back when they were still working actively.

AAAAH now I see. Honestly I've been always been a bit confused about these statuses. I'll fix it. So ONLY when it's active you want to deactivate?

**Comment by dnil** [comment]
I think so - can't think of any other.

**Comment by dnil** [comment]
I think it helps to consider that the inactive pile is very much treated as the INBOX of the institutes.

**Comment by dnil** [comment]
Inactive is not so much a "not to be worked on" as "alert - here is something that needs a worker asap"

**Comment by northwestwitch** [comment]
Fixed, I think

**Comment by northwestwitch** [comment]
But it might be the other way around actually

**Comment by dnil** [comment]
I still think harmonising with the web version - eg checking for other assigned users - would be good. If there is still an assigned user left after deletion, there is little reason to find the case among the "inactive" ones?

**Comment by northwestwitch** [comment]
> I still think harmonising with the web version - eg checking for other assigned users - would be good.

Yes, totally. I'll include a check, I was convinced that I did it, but it was only in my mind apparently..

**Comment by dnil** [comment]
It is somehow hard to think about since it works in the web interface. The intuition becomes that it must also already work in the CLI, although..

**Comment by northwestwitch** [comment]
Fixed for good, perhaps

**Comment by dnil** [comment]
Nice! I'll have a better look later - probably better tomorrow morning..

File: scout/commands/delete/delete_command.py [hunk 1]
```diff
@@ -63,7 +63,10 @@ def user(mail):
             link = url_for(
                 "cases.case", institute_id=institute_obj["_id"], case_name=case_obj["display_name"]
             )
-            if adapter.unassign(institute_obj, case_obj, user_obj, link, True):
+            inactivate_case = case_obj.get("status", "active") == "active" and case_obj[
```
Comment (dnil) [inline scout/commands/delete/delete_command.py:66]: 👍

File: scout/commands/delete/delete_command.py [hunk 2]
```diff
@@ -63,7 +63,10 @@ def user(mail):
...
+                "assignees"
+            ] == [mail]
```
Comment (dnil) [inline scout/commands/delete/delete_command.py:68]: 👍

File: tests/commands/delete/test_delete_cmd.py [hunk 3]
```diff
@@ -93,8 +93,85 @@ def test_delete_nonexisting_user(empty_mock_app, user_obj):
     assert "User unknown_email@email.com could not be found in database" in result.output
 
 
-def test_delete_user(empty_mock_app, user_obj, case_obj, institute_obj):
-    "Test the CLI command that will delete a user"
+def test_delete_last_user_active_case(empty_mock_app, user_obj, case_obj, institute_obj):
```
Comment (dnil) [inline tests/commands/delete/test_delete_cmd.py:96]: 👍

File: tests/commands/delete/test_delete_cmd.py [hunk 4]
```diff
@@ -93,8 +93,85 @@ def test_delete_nonexisting_user(empty_mock_app, user_obj):
...
+    assert store.case_collection.find_one({"assignees": {"$in": case_obj["assignees"]}})
+
+    ## WHEN deleting the user from the CLI
+    result = runner.invoke(cli, ["delete", "user", "-m", user_obj["email"]])
+
+    ## THEN the user should be gone
+    assert result.exit_code == 0
+    assert store.user_collection.find_one() is None
+
+    # The case should NOT be inactivated and should still have the other assignee
```
Comment (dnil) [inline tests/commands/delete/test_delete_cmd.py:164]: 👍

File: tests/commands/delete/test_delete_cmd.py [hunk 5]
```diff
@@ -121,8 +199,11 @@ def test_delete_user(empty_mock_app, user_obj, case_obj, institute_obj):
     assert result.exit_code == 0
     assert store.user_collection.find_one() is None
 
-    ## The case should not have an assignee
-    assert store.case_collection.find_one({"assignees": {"$in": case_obj["assignees"]}}) is None
+    # The case should keep the original status and should not have an assignee
+    updated_case = store.case_collection.find_one({"_id": case_obj["_id"]})
+    assert updated_case["status"] == "solved"
```
Comment (dnil) [inline tests/commands/delete/test_delete_cmd.py:204]: 👍

**Review by dnil — APPROVED** [review]
Looks good to me! I'm tempted to say it is all covered by the tests, but I guess we have to check it. Did you try it on stage?

**Comment by northwestwitch** [comment]
> Looks good to me! I'm tempted to say it is all covered by the tests, but I guess we have to check it. Did you try it on stage?

No, just locally, but I guess can be tested easily on hasta stage. I can assign myself to some solved case and then remove my user. I'll test it!

**Comment by northwestwitch** [comment]
- Assigned myself to this case: https://scout-stage.scilifelab.se/cust000/F0007437-mip7, that is solved.

image
 - Installed this branch on hasta cli stage

Removed myself from cases where I was assigned:

image

I was assigned on 5 cases : 
F0007437-mip7, F0009130-mip7, F0000496, 15035-validation, 18069

- Added again my user otherwise I can't login.

- Case is still solved but I'm not an assignee any more:

image

**Comment by northwestwitch** [comment]
Looks like it's working. Merging!

**Pull request merged**